lib/srv/usermgmt_test.go

```diff
@@ -417,35 +418,139 @@ func TestUpdateUserGroups(t *testing.T) {
+	assert.ElementsMatch(t, append(userinfo.Groups, types.TeleportManagedGroup), backend.users["alice"])
 
 	// Upsert again with same groups should not call SetUserGroups.
 	closer, err = users.UpsertUser("alice", userinfo)
 	assert.NoError(t, err)
 	assert.Equal(t, nil, closer)
 	assert.Equal(t, 1, backend.setUserGroupsCalls)
-	assert.ElementsMatch(t, userinfo.Groups, backend.users["alice"])
+	assert.ElementsMatch(t, append(userinfo.Groups, types.TeleportManagedGroup), backend.users["alice"])
+
+	// Updates with INSECURE_DROP mode should convert the managed user
+	userinfo.Mode = types.CreateHostUserMode_HOST_USER_MODE_INSECURE_DROP
+	userinfo.Groups = slices.Clone(allGroups[:2])
+	closer, err = users.UpsertUser("alice", userinfo)
+	assert.NoError(t, err)
+	assert.NotEqual(t, nil, closer)
+	assert.Equal(t, 2, backend.setUserGroupsCalls)
+	assert.ElementsMatch(t, append(userinfo.Groups, types.TeleportServiceGroup), backend.users["alice"])
+}
+
+func Test_UpdateUserGroups_Drop(t *testing.T) {
+	t.Parallel()
+	backend := newTestUserMgmt()
+	bk, err := memory.New(memory.Config{})
+	require.NoError(t, err)
+	pres := local.NewPresenceService(bk)
+	users := HostUserManagement{
+		backend: backend,
+		storage: pres,
+	}
+
+	allGroups := []string{"foo", "bar", "baz", "quux"}
+	for _, group := range allGroups {
+		require.NoError(t, backend.CreateGroup(group, ""))
+	}
+
+	userinfo := services.HostUsersInfo{
+		Groups: slices.Clone(allGroups[:2]),
+		Mode:   types.CreateHostUserMode_HOST_USER_MODE_INSECURE_DROP,
+	}
+
+	// Create user
+	closer, err := users.UpsertUser("alice", userinfo)
+	assert.NoError(t, err)
+	assert.NotEqual(t, nil, closer)
+	assert.Zero(t, backend.setUserGroupsCalls)
+	assert.ElementsMatch(t, append(userinfo.Groups, types.TeleportServiceGroup), backend.users["alice"])
+
+	// Update user with new groups.
+	userinfo.Groups = slices.Clone(allGroups[2:])
+
+	closer, err = users.UpsertUser("alice", userinfo)
+	assert.NoError(t, err)
+	assert.NotEqual(t, nil, closer)
+	assert.Equal(t, 1, backend.setUserGroupsCalls)
+	assert.ElementsMatch(t, append(userinfo.Groups, types.TeleportServiceGroup), backend.users["alice"])
+
+	// Upsert again with same groups should not call SetUserGroups.
+	closer, err = users.UpsertUser("alice", userinfo)
+	assert.NoError(t, err)
+	assert.NotEqual(t, nil, closer)
+	assert.Equal(t, 1, backend.setUserGroupsCalls)
+	assert.ElementsMatch(t, append(userinfo.Groups, types.TeleportServiceGroup), backend.users["alice"])
+
+	// Updates with KEEP mode should convert the ephemeral user
+	userinfo.Mode = types.CreateHostUserMode_HOST_USER_MODE_KEEP
+	userinfo.Groups = slices.Clone(allGroups[:2])
+	closer, err = users.UpsertUser("alice", userinfo)
+	assert.NoError(t, err)
+	assert.Equal(t, nil, closer)
+	assert.Equal(t, 2, backend.setUserGroupsCalls)
+	assert.Equal(t, 1, backend.createHomeDirectoryCalls)
+	assert.ElementsMatch(t, append(userinfo.Groups, types.TeleportManagedGroup), backend.users["alice"])
```

Comment:
I actually had this originally and then dropped it since the `ElementsMatch` _technically_ covers that case already. I think the explicit check helps better convey what we're asserting, though, and gives us some extra safety in case there's a future bug where `userinfo.Groups` gets mutated :+1: